### PR TITLE
[storage/index] Add More Test Coverage + Fix `None` Panic with Full State Machine Refactor

### PR DESCRIPTION
**Pull request by patrick-ogrady**
Resolves: #886 

## TODO

- [x] Add metric for `translated_keys` and `items` (all items, not just map)

### CODE REVIEW

File: storage/src/index/storage.rs [hunk 1]
```diff
@@ -60,6 +67,7 @@ pub struct Cursor<'a, T: Translator, V> {
     next: Option<Box<Record<V>>>,
     past: Option<Box<Record<V>>>,
 
+    next_required: bool,
```
Comment (patrick-ogrady) [inline storage/src/index/storage.rs]: This fixes an issue where you could call `insert(), insert()` or `insert(), delete()`.

File: storage/src/index/storage.rs [hunk 2]
```diff
@@ -201,6 +226,11 @@ impl<'a, T: Translator, V> Cursor<'a, T, V> {
 
                 // Set current next to be the new record.
                 self.next = Some(new);
+
+                // Now that the current item is in next, we need to update
+                // the phase.
+                self.phase = Phase::Next;
```
Comment (patrick-ogrady) [inline storage/src/index/storage.rs]: This was also a bug (where adding an item would stay at `Entry`).
